### PR TITLE
Port pyximport to importlib machinery with git history

### LINKED ISSUE #4560: [BUG] pyximport uses deprecated "imp" module
Replace it with `importlib` or whatever seems appropriate and modern.

### PR DESCRIPTION
**Pull request by matusvalo**
This is another attempt for migrating pyximport to importlib machinery. This attempt is based on #4612 but with hope that history is kept in both _pyximport2.py and _pyximport3.py files.

> *Note*: Both _pyximport2.py and _pyximport3.py file should contain git history of original pyximport.py file. But pyximport.py file is newly created and does not contain history of previous pyximport.py. I don't think it is problem since both _pyximport2.py and _pyximport3.py files should have this history.

Fixes #4560

### CODE REVIEW

**Comment by matusvalo** [comment]
@scoder @da-woods feel free to close #4612 if you are fine with handling of GIT history in this PR.

**Comment by mattip** [comment]
When I do `git log --follow pyximport/_pyximport2.py` or `git log --follow pyximport/_pyximport3.py`, I see the complete log. I think this is as good as it gets.

**Comment by mattip** [comment]
I think this is ready for a review

**Comment by mattip** [comment]
Ping. This PR is a prerequisite for #4630, which is needed for windows.

**Comment by mattip** [comment]
What is the path forward for this PR? Is there some other approach to the change that will make it easier to review and merge?

**Comment by matusvalo** [comment]
> What is the path forward for this PR? Is there some other approach to the change that will make it easier to review and merge?

I think we should bring this to cython-devel mailing list.

**Comment by mattip** [comment]
@scoder thoughts?

File: pyximport/pyximport.py [hunk 1]
```diff
@@ -1,606 +1,11 @@
...
-            setattr(item, "__module__", mod_name)
-        except (AttributeError, TypeError):
-            pass
-    help(__main__)
 
+if sys.version_info.major == 3 and sys.version_info.minor >= 5:
+    # _pyximport3 module requires at least Python 3.5
+    from pyximport._pyximport3 import install, uninstall, show_docs
+else:
+    from pyximport._pyximport2 import install, uninstall, show_docs
```
Comment (scoder) [inline pyximport/pyximport.py:8]: Let's not write code that we already know will break in Python 4.
```suggestion
if sys.version_info < (3, 5):
    # _pyximport3 module requires at least Python 3.5
    from pyximport._pyximport2 import install, uninstall, show_docs
else:
    from pyximport._pyximport3 import install, uninstall, show_docs
```

File: pyximport/_pyximport3.py [hunk 2]
```diff
@@ -205,283 +213,142 @@ def build_module(name, pyxfilename, pyxbuild_dir=None, inplace=False, language_l
...
+        try:
+            return super().exec_module(module)
+        except Exception as failure_exc:
+            import traceback
+            _debug("Failed to load extension module: %r" % failure_exc)
+            raise ImportError("Executing module %s failed %s" % (
+                    module.__file__, traceback.format_exception_only(*sys.exc_info()[:2])))
 
 
-#install args
```
Comment (scoder) [inline pyximport/_pyximport3.py]: Why remove the comment?

File: pyximport/_pyximport3.py [hunk 3]
```diff
@@ -205,283 +213,142 @@ def build_module(name, pyxfilename, pyxbuild_dir=None, inplace=False, language_l
...
-                                 language_level=self.language_level)
-        return module
+    def exec_module(self, module):
+        try:
+            return super().exec_module(module)
+        except Exception as failure_exc:
+            import traceback
+            _debug("Failed to load extension module: %r" % failure_exc)
+            raise ImportError("Executing module %s failed %s" % (
+                    module.__file__, traceback.format_exception_only(*sys.exc_info()[:2])))
```
Comment (scoder) [inline pyximport/_pyximport3.py:337]: Do we have tests for these failure cases?

File: pyximport/_pyximport3.py [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+class PyxImportMetaFinder(MetaPathFinder):
+
+    def __init__(self, extension=PYX_EXT, pyxbuild_dir=None, inplace=False, language_level=None):
+        self.pyxbuild_dir = pyxbuild_dir
+        self.inplace = inplace
+        self.language_level = language_level
+        self.extension = extension
+
+    def find_spec(self, fullname, path, target=None):
+        if path is None or path == "":
```
Comment (scoder) [inline pyximport/_pyximport3.py:227]:
```suggestion
        if not path:
```

File: pyximport/_pyximport3.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+            return None
+        if any([fullname.startswith(pkg) for pkg in self.blocked_packages]):
+            return None
+        if fullname in self.blocked_modules:
+            # prevent infinite recursion
+            return None
+
+        self.blocked_modules.append(fullname)
+        name = fullname
+        if path is None or path == "":
```
Comment (scoder) [inline pyximport/_pyximport3.py:275]:
```suggestion
        if not path:
```

File: pyximport/_pyximport3.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+            path = [os.getcwd()]  # top level import --
+        if "." in fullname:
+            *parents, name = fullname.split(".")
+        else:
+            name = fullname
+        for entry in path:
+            if os.path.isdir(os.path.join(entry, name)):
+                # this module has child modules
+                filename = os.path.join(entry, name, "__init__.py")
```
Comment (scoder) [inline pyximport/_pyximport3.py:236]:
```suggestion
                filename = os.path.join(entry, name, "__init__" + self.extension)
```

File: pyximport/_pyximport3.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+        self.blocked_modules.append(fullname)
+        name = fullname
+        if path is None or path == "":
+            path = [os.getcwd()]  # top level import --
+        try:
+            for entry in path:
+                if os.path.isdir(os.path.join(entry, name)):
+                    # this module has child modules
+                    filename = os.path.join(entry, name, "__init__.py")
```
Comment (scoder) [inline pyximport/_pyximport3.py:281]:
```suggestion
                    filename = os.path.join(entry, name, "__init__" + self.extension)
```

File: pyximport/_pyximport3.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+                    submodule_locations = [os.path.join(entry, name)]
+                else:
+                    filename = os.path.join(entry, name + PY_EXT)
```
Comment (scoder) [inline pyximport/_pyximport3.py:284]:
```suggestion
                    filename = os.path.join(entry, name + self.extension)
```

File: pyximport/_pyximport3.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+
+# Performance problem: for every PYX file that is imported, we will
+# invoke the whole distutils infrastructure even if the module is
+# already built. It might be more efficient to only do it when the
+# mod time of the .pyx is newer than the mod time of the .so but
+# the question is how to get distutils to tell me the name of the .so
+# before it builds it. Maybe it is easy...but maybe the performance
+# issue isn't real.
+def _load_pyrex(name, filename):
+    "Load a pyrex file given a name and filename."
```
Comment (scoder) [inline pyximport/_pyximport3.py:97]: Why is this function empty?

**Review by scoder — COMMENTED** [review]
Sorry for the really long delay, and thanks, @matusvalo.
I added some comments.

File: pyximport/_pyximport3.py (hunk 2 above)
Comment (matusvalo) [reply pyximport/_pyximport3.py]: By mistake. Fixed.

File: pyximport/_pyximport3.py (hunk 9 above)
Comment (matusvalo) [reply pyximport/_pyximport3.py:97]: It is empty also in original implementation. I did not see this function used anywhere. Can this function be removed? (I ran unittests related to pyximport and they seems to be passing with this function is removed.)

Comment (da-woods) [reply pyximport/_pyximport3.py:97]: Looking at the git history I'm not sure it was ever used or that it ever did anything. So remove it.

File: pyximport/_pyximport3.py (hunk 3 above)
Comment (matusvalo) [reply pyximport/_pyximport3.py:337]: Unfortunately not. How this failure can be implemented? `exec_module()` internally uses low level machinery which I cannot "break":
https://github.com/python/cpython/blob/6217864fe5f6855f59d608733ce83fd4466e1b8c/Lib/importlib/_bootstrap_external.py#L1236-L1240
Corrupting .so file causes segmentation fault. Is it possible to use mock for this test?

File: pyximport/_pyximport3.py (hunk 9 above)
Comment (matusvalo) [reply pyximport/_pyximport3.py]: Removed.

**Comment by scoder** [comment]
Thanks @matusvalo !

**Pull request merged**